src/System.Buffers.Primitives/System/Buffers/Buffer.cs

```diff
@@ -126,28 +126,38 @@ public Span<T> Span
             }
         }
 
-        public BufferHandle Retain()
+        public BufferHandle Retain(bool pin = false)
         {
-            if (_owner != null)
+            BufferHandle bufferHandle;
+            if (pin)
             {
-                _owner.Retain();
-                return new BufferHandle(_owner);
+                if (_owner != null)
+                {
+                    bufferHandle = _owner.Pin(_index);
+                }
+                else
+                {
+                    var handle = GCHandle.Alloc(_array, GCHandleType.Pinned);
+                    unsafe
+                    {
+                        var pointer = OwnedBuffer<T>.Add((void*)handle.AddrOfPinnedObject(), _index);
+                        bufferHandle = new BufferHandle(null, pointer, handle);
+                    }
+                }
             }
-            return new BufferHandle();
-        }
-
-        public BufferHandle Pin()
-        {
-            if (_owner != null)
+            else
             {
-                return _owner.Pin(_index);
-            }
-            var handle = GCHandle.Alloc(_array, GCHandleType.Pinned);
-            unsafe
-            {
-                var pointer = OwnedBuffer<T>.Add((void*)handle.AddrOfPinnedObject(), _index);
-                return new BufferHandle(null, pointer, handle);
+                if (_owner != null)
```

Comment:
Is that likely to change or is that the correct behaviour? @KrzysztofCwalina 

If using the default constructor of BufferHandle is acceptable (where void* _pointer == null and IRetainable _owner == null), then I will make this change.